### PR TITLE
liveness: introduce NodeVitality

### PR DESCRIPTION
**Pull request by andrewbaptist**
Introduce a new NodeVitality concept which is derived from liveness and
StoreDescriptor updates. It will be used everywhere other than in
epoch leases which depend directly on the liveness range.

Epic: none

Release note: None

### CODE REVIEW

**Review by erikgrinaker — COMMENTED** [review]
I think we should set up a call next week to sync up on what the desired end state here is. I thought we were going to move towards inconsistent liveness signals propagated via gossip and strongly consistent cluster membership stored in KV, but we seem to be conflating these concerns here. Flushing the comments I've written so far, but they're mostly variations on that theme.

> Reviewed 6 of 8 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103975#-:-NWSg5jA4RWrKL6SHoKg:b-1surib)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist)

___
*[`pkg/kv/kvserver/liveness/cache.go` line 214 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/103975#-NWSaO-j-KASc9K3G9Gn:-NWSaO-j-KASc9K3G9Go:b-gka0vm) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1fb04f90ce74c6751276866c1af6e38e0109e9b/pkg/kv/kvserver/liveness/cache.go#L214)):*
> ```Go
> }
> 
> func (c *cache) GetActiveNodes() []roachpb.NodeID {
> ```

If this is about cluster membership, shouldn't we read that from the store? I thought the idea here was to use the cache for liveness info via gossip, but KV for cluster membership. Would appreciate if you could clarify the intended separation of concerns between these concepts and actors.
___
*[`pkg/kv/kvserver/liveness/cache.go` line 226 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/103975#-NWSZwxVAbvFqPwpVDe7:-NWSZwxVAbvFqPwpVDe8:b-vvdlim) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1fb04f90ce74c6751276866c1af6e38e0109e9b/pkg/kv/kvserver/liveness/cache.go#L226)):*
> ```Go
> }
> 
> func (c *cache) LastDescriptorUpdate(nodeID roachpb.NodeID) (hlc.Timestamp, bool) {
> ```

Let's be explicit about what kind of descriptors these are. I'm finding it pretty confusing that we imply these are node descriptors but the updates are for store descriptors. Is there any value in considering liveness on a per-store basis independently from node liveness? To illustrate my confusion here, we have `TimeUntilStoreDead`, which seems to apply to nodes more than stores, but should it also apply individually to stores? If not, why don't we call it `TimeUntilNodeDead`?
___
*[`pkg/kv/kvserver/liveness/liveness.go` line 984 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/103975#-NWSfsg-0GnuXnzEeP0X:-NWSfsg-0GnuXnzEeP0Y:b-13mi4k) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1fb04f90ce74c6751276866c1af6e38e0109e9b/pkg/kv/kvserver/liveness/liveness.go#L984)):*
> ```Go
> 
> // GetNodeStatusFromKV returns the status for all the nodes from KV.
> func (nl *NodeLiveness) GetNodeStatusFromKV(
> ```

Wasn't the idea to only fetch liveness info from gossip? Why are we reading it from KV?
___
*[`pkg/kv/kvserver/liveness/livenesspb/liveness.go` line 156 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/103975#-NWSXxis2uuqQ5spqsLJ:-NWSXxitBENASjzz_NDe:bfkksjh) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1fb04f90ce74c6751276866c1af6e38e0109e9b/pkg/kv/kvserver/liveness/livenesspb/liveness.go#L156)):*
> ```Go
> // NodeStatus should be used any place other than epoch leases where it is
> // necessary to determine if a node is currently alive and what its health is.
> type NodeStatus struct {
> ```

`NodeStatus` is an unfortunate name, because it's already used by a different concept. I'd recommend we come up with some clear, unambiguous terms here.

https://github.com/cockroachdb/cockroach/blob/a9989a85485c1796fd11907ae2cc019c962c1818/pkg/keys/keys.go#L179-L186

https://github.com/cockroachdb/cockroach/blob/e890c334ceb108fbac8d07af43426337fbae46ea/pkg/server/status/statuspb/status.proto#L28-L30

https://github.com/cockroachdb/cockroach/blob/62c6518f1ac602dcdc1557fd2c5b0f54147a9a4b/pkg/sql/crdb_internal.go#L5071-L5093
___
*[`pkg/kv/kvserver/liveness/livenesspb/liveness.go` line 159 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/103975#-NWSccsy60w85y2dIhzd:-NWSccsy60w85y2dIhze:btugm4f) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1fb04f90ce74c6751276866c1af6e38e0109e9b/pkg/kv/kvserver/liveness/livenesspb/liveness.go#L159)):*
> ```Go
> 	draining   bool
> 	membership MembershipStatus
> 	isDead     bool
> ```

These various fields should have comments explaining exactly what they represent, and their conditions.
___
*[`pkg/kv/kvserver/liveness/livenesspb/liveness.go` line 173 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/103975#-NWSXdQc1LoGo2u7Ylp6:-NWSXdQc1LoGo2u7Ylp7:b-iktup0) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1fb04f90ce74c6751276866c1af6e38e0109e9b/pkg/kv/kvserver/liveness/livenesspb/liveness.go#L173)):*
> ```Go
> // used when the caller needs to be able to contact leaseholders directly.
> // Returns false if the node is not in the local liveness table.
> func (ne NodeStatus) IsAvailableNotDraining() bool {
> ```

nit: `ne`?
___
*[`pkg/kv/kvserver/liveness/livenesspb/liveness.go` line 217 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/103975#-NWSbUwt0GktMdiecwb3:-NWSbUwt0GktMdiecwb4:b-nmwcq2) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1fb04f90ce74c6751276866c1af6e38e0109e9b/pkg/kv/kvserver/liveness/livenesspb/liveness.go#L217)):*
> ```Go
> // ideally we should remove usage of NodeLivenessStatus altogether. See #50707
> // for more details.
> func (ne NodeStatus) LivenessStatus() NodeLivenessStatus {
> ```

Is this just temporary until we can remove it? I think we've wanted to get away from this for a long time. It's a historical vestige from before we had explicit membership states, and as the comment points out, it differs from the canonical membership state. I also thought we wanted to separate membership from liveness. If not, then we should make sure this agrees with the canonical membership state.
___
*[`pkg/kv/kvserver/liveness/livenesspb/liveness.go` line 273 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/103975#-NWSeKHiEkW9IOSFKz6p:-NWSeKHiEkW9IOSFKz6q:b-e5n69j) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1fb04f90ce74c6751276866c1af6e38e0109e9b/pkg/kv/kvserver/liveness/livenesspb/liveness.go#L273)):*
> ```Go
> }
> 
> // TestRestartNode increments the epoch for a given node and marks it as
> ```

nit: epoch mention seems outdated?
___
*[`pkg/kv/kvserver/liveness/client_test.go` line 335 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/103975#-NWSRWzp2fT073p2oyiE:-NWSRWzp2fT073p2oyiF:bloocl1) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1fb04f90ce74c6751276866c1af6e38e0109e9b/pkg/kv/kvserver/liveness/client_test.go#L335)):*
> ```Go
> // TestNodeLivenessNodeCount tests GetNodeCount() and GetNodeCountWithOverrides,
> // which are critical for computing the number of needed voters for a range.
> func TestNodeLivenessNodeCount(t *testing.T) {
> ```

Rename this to `TestGetActiveNodes` and assert on the returned node IDs too.
___
*[`pkg/kv/kvserver/liveness/client_test.go` line 379 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/103975#-NWSRA3C0FJ5VVq4L-NC:-NWSRA3C0FJ5VVq4L-ND:blbazn2) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1fb04f90ce74c6751276866c1af6e38e0109e9b/pkg/kv/kvserver/liveness/client_test.go#L379)):*
> ```Go
> 		5: livenesspb.NodeLivenessStatus_DECOMMISSIONING,
> 	}
> 	require.Equal(t, numNodes, storepool.GetNodeCountWithOverrides(nl1, nil))
> ```

These test cases should be pulled out to a separate test for `GetNodeCountWithOverrides` in the `storepool` package.

**Review by andrewbaptist — COMMENTED** [review]
I set something up for tomorrow. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103975#-:-NWiBQRBDu8718EFc3WX:b-7drhe5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker)

___
*[`pkg/kv/kvserver/liveness/cache.go` line 214 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/103975#-NWSaO-j-KASc9K3G9Gn:-NWiMqk4AOEWrm_MM49X:b82uaek) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1fb04f90ce74c6751276866c1af6e38e0109e9b/pkg/kv/kvserver/liveness/cache.go#L214)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

If this is about cluster membership, shouldn't we read that from the store? I thought the idea here was to use the cache for liveness info via gossip, but KV for cluster membership. Would appreciate if you could clarify the intended separation of concerns between these concepts and actors.
</blockquote></details>

There are a few PRs here that move in this direction. This PR is specifically keeping the existing behavior for most of the code to avoid changing tests, but still making it so that we are using the new interfaces to do it. In terms of reading from the cache or the store, there is an overlap period when we still have to support epoch leases, and the methods here are still mixed between them.
___
*[`pkg/kv/kvserver/liveness/cache.go` line 226 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/103975#-NWSZwxVAbvFqPwpVDe7:-NWiO5ZzD-RHeoe6CoYB:bspfpfd) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1fb04f90ce74c6751276866c1af6e38e0109e9b/pkg/kv/kvserver/liveness/cache.go#L226)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Let's be explicit about what kind of descriptors these are. I'm finding it pretty confusing that we imply these are node descriptors but the updates are for store descriptors. Is there any value in considering liveness on a per-store basis independently from node liveness? To illustrate my confusion here, we have `TimeUntilStoreDead`, which seems to apply to nodes more than stores, but should it also apply individually to stores? If not, why don't we call it `TimeUntilNodeDead`?
</blockquote></details>

Updated (in a separate commit, same PR)
___
*[`pkg/kv/kvserver/liveness/liveness.go` line 984 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/103975#-NWSfsg-0GnuXnzEeP0X:-NWiOAEI42BHkKc5xihu:bpegc1s) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1fb04f90ce74c6751276866c1af6e38e0109e9b/pkg/kv/kvserver/liveness/liveness.go#L984)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Wasn't the idea to only fetch liveness info from gossip? Why are we reading it from KV?
</blockquote></details>

The existing code has a mix of reading from gossip and KV. Some places want the KV data because of gossip timing races (they update something and want to see the result immediately). Also the admin interfaces typically read from KV directly. I am not adding any new uses of reading from KV, just maintaining the existing ones.
___
*[`pkg/kv/kvserver/liveness/livenesspb/liveness.go` line 156 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/103975#-NWSXxis2uuqQ5spqsLJ:-NWiOeCT-xAnyu4tC_p0:b-q82kva) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1fb04f90ce74c6751276866c1af6e38e0109e9b/pkg/kv/kvserver/liveness/livenesspb/liveness.go#L156)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

`NodeStatus` is an unfortunate name, because it's already used by a different concept. I'd recommend we come up with some clear, unambiguous terms here.

https://github.com/cockroachdb/cockroach/blob/a9989a85485c1796fd11907ae2cc019c962c1818/pkg/keys/keys.go#L179-L186

https://github.com/cockroachdb/cockroach/blob/e890c334ceb108fbac8d07af43426337fbae46ea/pkg/server/status/statuspb/status.proto#L28-L30

https://github.com/cockroachdb/cockroach/blob/62c6518f1ac602dcdc1557fd2c5b0f54147a9a4b/pkg/sql/crdb_internal.go#L5071-L5093
</blockquote></details>

Good point. I renamed to NodeVitality.
___
*[`pkg/kv/kvserver/liveness/livenesspb/liveness.go` line 159 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/103975#-NWSccsy60w85y2dIhzd:-NWiSNwQ0mhHmu0dk1bY:b7qm0r6) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1fb04f90ce74c6751276866c1af6e38e0109e9b/pkg/kv/kvserver/liveness/livenesspb/liveness.go#L159)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

These various fields should have comments explaining exactly what they represent, and their conditions.
</blockquote></details>

I added comments about this condition.
___
*[`pkg/kv/kvserver/liveness/livenesspb/liveness.go` line 173 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/103975#-NWSXdQc1LoGo2u7Ylp6:-NWiSLX66oWBmRuYxGKW:b88p6zs) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1fb04f90ce74c6751276866c1af6e38e0109e9b/pkg/kv/kvserver/liveness/livenesspb/liveness.go#L173)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

nit: `ne`?
</blockquote></details>

Updated this to nv (NodeVitality)
___
*[`pkg/kv/kvserver/liveness/livenesspb/liveness.go` line 217 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/103975#-NWSbUwt0GktMdiecwb3:-NWiSzjcC-V091-w8SIb:b-7phl6c) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1fb04f90ce74c6751276866c1af6e38e0109e9b/pkg/kv/kvserver/liveness/livenesspb/liveness.go#L217)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Is this just temporary until we can remove it? I think we've wanted to get away from this for a long time. It's a historical vestige from before we had explicit membership states, and as the comment points out, it differs from the canonical membership state. I also thought we wanted to separate membership from liveness. If not, then we should make sure this agrees with the canonical membership state.
</blockquote></details>

Yes, this is temporary for now. I will add a comment that this should be removed. I will fold it into NodeVitality in a later PR.
___
*[`pkg/kv/kvserver/liveness/livenesspb/liveness.go` line 273 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/103975#-NWSeKHiEkW9IOSFKz6p:-NWiTVhtE2IyDdAqvQ2e:bq5wuc9) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1fb04f90ce74c6751276866c1af6e38e0109e9b/pkg/kv/kvserver/liveness/livenesspb/liveness.go#L273)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

nit: epoch mention seems outdated?
</blockquote></details>

Updated comment.
___
*[`pkg/kv/kvserver/liveness/client_test.go` line 335 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/103975#-NWSRWzp2fT073p2oyiE:-NWiU-2433x-iwqzkoCC:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1fb04f90ce74c6751276866c1af6e38e0109e9b/pkg/kv/kvserver/liveness/client_test.go#L335)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Rename this to `TestGetActiveNodes` and assert on the returned node IDs too.
</blockquote></details>

Done
___
*[`pkg/kv/kvserver/liveness/client_test.go` line 379 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/103975#-NWSRA3C0FJ5VVq4L-NC:-NWii0dCCeOsxlh0Gu7m:b-isfset) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1fb04f90ce74c6751276866c1af6e38e0109e9b/pkg/kv/kvserver/liveness/client_test.go#L379)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

These test cases should be pulled out to a separate test for `GetNodeCountWithOverrides` in the `storepool` package.
</blockquote></details>

This was unfortunately a little difficult to do. I left a comment on them for now and will try and move them in a separate commit.

**Review by erikgrinaker — APPROVED** [review]
Discussed the overall approach elsewhere. For now, let's just start collecting liveness logic in one place, and we can iterate on it. I'm assuming you've verified there are no significant logic changes here, so I haven't reviewed that in detail -- beyond that, I think we've done enough bikeshedding here. :) Consider getting another review from someone more actively involved in this area.

> Reviewed 10 of 27 files at r4, 11 of 11 files at r6, 8 of 8 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103975#-:-NXVih1I5iHP6l1BVuqd:bd8jogt)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist)

___
*[`pkg/kv/kvserver/liveness/cache.go` line 214 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/103975#-NWSaO-j-KASc9K3G9Gn:-NXVax1-DWRnScan2XFO:b6je8hy) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1fb04f90ce74c6751276866c1af6e38e0109e9b/pkg/kv/kvserver/liveness/cache.go#L214)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

There are a few PRs here that move in this direction. This PR is specifically keeping the existing behavior for most of the code to avoid changing tests, but still making it so that we are using the new interfaces to do it. In terms of reading from the cache or the store, there is an overlap period when we still have to support epoch leases, and the methods here are still mixed between them.
</blockquote></details>

Discussed elsewhere.
___
*[`pkg/kv/kvserver/liveness/cache.go` line 226 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/103975#-NWSZwxVAbvFqPwpVDe7:-NXVU5TZ0bKooXRWPl9N:bbyt645) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1fb04f90ce74c6751276866c1af6e38e0109e9b/pkg/kv/kvserver/liveness/cache.go#L226)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

Updated (in a separate commit, same PR)
</blockquote></details>

Well, this wasn't so much about renaming `TimeUntilStoreDead` (which can be confusing considering we probably don't want to rename the `server.time_until_store_dead` setting). I was wondering more generally about whether we should be considering individual store health, or if we only care about node health.

**Review by andrewbaptist — COMMENTED** [review]
Correct this PR doesn't really "change" anything. It just adds methods that are used by future PRs (which do change things, but they are all small independent changes). I chatted in the KV team this week about the changes, and there was general agreement that this is the right direction. There was some idea of "changing the way" that the "IsAlive" is done (something more like "IsAliveForUsage(Usage)" - and then centralizing the mapping from "Usage -> what is considered alive" in a single place. I liked that idea, but that will be an easy "add-on" before the rest of the PRs merge. We can chat on Monday on your thoughts on this.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/103975#-:-NXWeAWD0XYobyz_LWc_:b-h09gcg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker)

___
*[`pkg/kv/kvserver/liveness/cache.go` line 226 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/103975#-NWSZwxVAbvFqPwpVDe7:-NXWeglC7VLsf1oFxVAY:b-ssnaau) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1fb04f90ce74c6751276866c1af6e38e0109e9b/pkg/kv/kvserver/liveness/cache.go#L226)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Well, this wasn't so much about renaming `TimeUntilStoreDead` (which can be confusing considering we probably don't want to rename the `server.time_until_store_dead` setting). I was wondering more generally about whether we should be considering individual store health, or if we only care about node health.
</blockquote></details>

I see what you mean. I don't think we want to try and consider individual store health from a "dead/alive" perspective at this point. It is a little confusing about what it means for things other than connectivity, but it does seem wrong to consider a store within a node dead while the other stores are alive. I'll make sure this is consistently named for now, and not change `server.time_until_store_dead` since that might require an upgrade (although its unlikely anyone touched that setting).

**Comment by andrewbaptist** [comment]
I'm going to hold off merging until I investigate more. This change seems to induce flakiness in some of the TestTenantLogic tests. I can't tell if the tests are written poorly or if this change is causing a problem.

**Comment by andrewbaptist** [comment]
bors r=erikgrinaker

**Comment by andrewbaptist** [comment]
bors retry

**Pull request merged**
